src/transformStateWithClones.js

```diff
@@ -7,7 +9,30 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  const history = [];
+  let current = { ...state };
+
+  for (const action of actions) {
+    let newCurrent = { ...current };
+
+    switch (action.type) {
+      case 'addProperties':
+        newCurrent = { ...current, ...action.extraData };
+        break;
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete newCurrent[key];
+        }
+        break;
+      case 'clear':
+        newCurrent = {};
+        break;
+    }
```

Comment:
The switch statement correctly handles the action types. Ensure that the `action` object always contains the expected properties (`extraData`, `keysToRemove`) to avoid runtime errors.